includes/plugins/co-authors-plus/class-guest-author-role.php

```diff
@@ -0,0 +1,277 @@
+		}
+
+		if ( ! empty( $errors->errors['user_login'] ) ) {
+			$errors->remove( 'user_login' );
+		}
+
+		// We still don't want users with duplicate emails.
+		if ( ! empty( $errors->errors['email_exists'] ) ) {
+			return $errors;
+		}
+
+		if ( ! $update ) {
+			// For guest authors, the form is modified via JS and we get the display name in the username field.
+			$user->display_name = $user->user_login;
+
+			// Create user name from Display name.
+			$user->user_login = self::generate_username( $user->display_name );
+		}
+
+		return $errors;
+	}
+
+	/**
+	 * Generates a unique username from a display name.
+	 *
+	 * @param string $display_name The user's display name.
+	 * @return string
+	 */
+	public static function generate_username( $display_name ) {
+		$username = \sanitize_user( $display_name, true );
+		$username = \sanitize_title( $username );
+
+		while ( \username_exists( $username ) ) {
+			$username = $username . '-' . \wp_rand( 1, 100 );
+		}
+
+		return $username;
+	}
+
+	/**
+	 * Enqueues the JS that modifies the user profile and user creation forms.
+	 *
+	 * @return void
+	 */
+	public static function admin_footer() {
+		global $pagenow;
+		\wp_enqueue_script(
+			'newspack-co-authors-plus',
+			Newspack::plugin_url() . '/dist/other-scripts/co-authors-plus.js',
+			[ 'jquery' ],
+			NEWSPACK_PLUGIN_VERSION,
+			true
+		);
+
+		wp_localize_script(
+			'newspack-co-authors-plus',
+			'guestAuthorRole',
+			[
+				'role'             => self::CONTRIBUTOR_NO_EDIT_ROLE_NAME,
+				'displayNameLabel' => __( 'Display name', 'newspack-plugin' ),
+				'screen'           => $pagenow === 'user-new.php' ? 'new' : 'edit',
+			]
+		);
+	}
+
+	/**
+	 * A generic callback applied to filters that check if a user has access to a feature, or if a certain field should be displayed in its profile.
+	 *
+	 * These callbacks pass the return of the check as the first argument ant the user or user ID as the second.
+	 *
+	 * @param bool        $result The result of the check.
+	 * @param int|WP_User $user A user ID or user object.
+	 * @return bool
+	 */
+	public static function disable_feature( $result, $user ) {
+		if ( is_int( $user ) ) {
+			$user = \get_user_by( 'id', $user );
+		}
+
+		if ( ! is_a( $user, 'WP_User' ) ) {
+			return $result;
+		}
+
+		if ( in_array( self::CONTRIBUTOR_NO_EDIT_ROLE_NAME, $user->roles, true ) ) {
+			return false;
+		}
+
+		return $result;
+	}
+
+	/**
+	 * Filters user authentication to prevent guest authors from logging in.
+	 *
+	 * @param WP_Error|WP_User $user The logged in user or login error.
+	 * @param string           $password The user's password.
+	 * @return WP_Error|WP_User
+	 */
+	public static function wp_authenticate_user( $user, $password ) {
+		if ( ! is_a( $user, 'WP_User' ) ) {
+			return $user;
+		}
+
+		if ( in_array( self::CONTRIBUTOR_NO_EDIT_ROLE_NAME, $user->roles, true ) ) {
+			return new WP_Error( 'guest_authors_cannot_login', __( 'Guest authors cannot login.', 'newspack-plugin' ) );
```

Comment:
Why shouldn't these users be allowed to log in? The goal of the original implementation, in line with the requirements, was to allow _any_ user to become assignable to posts. Many sites have users who are subscribers (should be able to manage their subscriptions), but they've contributed to an article and are listed as authors. They should still be able to log in and manage their subscriptions. The only thing they should not be able to do is editing posts.

Reply:
We are blocking the users with this role from logging in, just as we are not event allowing a password to be set.

User with the custom capability that allows them to be assigned as co authors can still login. They are not affected.

Reply:
> User with the custom capability that allows them to be assigned as co authors can still login.

But users with multiple roles (e.g. `subscriber,contributor_no_edit`) won't be able to log in, and should.